src/torchmetrics/detection/iou.py

```diff
@@ -213,20 +216,36 @@ def _get_gt_classes(self) -> list:
 
     def compute(self) -> dict:
         """Computes IoU based on inputs passed in to ``update`` previously."""
-        score = torch.cat([mat[mat != self._invalid_val] for mat in self.iou_matrix], 0).mean()
+        # compute global IoU score using only valid values.
+        valid_matrices = [
+            mat[mat != self._invalid_val] for mat in self.iou_matrix if torch.any(mat != self._invalid_val)
+        ]
+        score = torch.cat(valid_matrices, 0).mean() if valid_matrices else torch.tensor(0.0)
         results: dict[str, Tensor] = {f"{self._iou_type}": score}
         if torch.isnan(score):  # if no valid boxes are found
             results[f"{self._iou_type}"] = torch.tensor(0.0, device=score.device)
         if self.class_metrics:
-            gt_labels = dim_zero_cat(self.groundtruth_labels)
-            classes = gt_labels.unique().tolist() if len(gt_labels) > 0 else []
+            # union of ground truth and predicted labels
+            all_labels = torch.tensor([], dtype=torch.long, device=score.device)
+            if self.groundtruth_labels:
+                all_labels = torch.cat(self.groundtruth_labels)
+            if self.pred_labels:
+                all_labels = torch.cat([all_labels, torch.cat(self.pred_labels)])
```

Comment:
```suggestion
            all_labels = dim_zero_cat([dim_zero_cat(self.groundtruth_labels), dim_zero_cat(self.pred_labels)])
```
this should work, tested locally